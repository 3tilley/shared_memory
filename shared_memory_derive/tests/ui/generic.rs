//! We should assert that each parameter type implements the trait and fail to
//! compile if it does not.

use shared_memory::SharedMemCast;

#[derive(SharedMemCast)]
pub struct MyStruct<T> {
    x: T,
}

#[derive(SharedMemCast)]
pub struct MyStruct2<T>(T);

#[derive(SharedMemCast)]
pub enum MyData<T> {
    Foo(i32),
    Bar {
        x: f64,
        yyy: [f64; 32],
        val: T,
    }
}

pub struct NotShared;

fn assert_impl<T: SharedMemCast>() {}
fn main() {
    assert_impl::<MyStruct<NotShared>>();
    assert_impl::<MyStruct<Vec<i32>>>();
    assert_impl::<MyStruct<&[i32]>>();
    fn generic_param<U>() {
        assert_impl::<MyStruct<U>>();
    }

    assert_impl::<MyStruct2<NotShared>>();
    assert_impl::<MyStruct2<Vec<i32>>>();
    assert_impl::<MyStruct2<&[i32]>>();
    fn generic_param2<U>() {
        assert_impl::<MyStruct2<U>>();
    }

    assert_impl::<MyData<NotShared>>();
    assert_impl::<MyData<Vec<i32>>>();
    assert_impl::<MyData<&[i32]>>();
    fn generic_param3<U>() {
        assert_impl::<MyData<U>>();
    }
}
