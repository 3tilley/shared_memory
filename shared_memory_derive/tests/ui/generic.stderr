error[E0277]: the trait bound `NotShared: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:28:5
   |
28 |     assert_impl::<MyStruct<NotShared>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `NotShared`
   |
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct<NotShared>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `std::vec::Vec<i32>: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:29:5
   |
29 |     assert_impl::<MyStruct<Vec<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `std::vec::Vec<i32>`
   |
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct<std::vec::Vec<i32>>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `&[i32]: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:30:5
   |
30 |     assert_impl::<MyStruct<&[i32]>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `&[i32]`
   |
   = help: the following implementations were found:
             <[T; 1024] as shared_memory::cast::SharedMemCast>
             <[T; 1048576] as shared_memory::cast::SharedMemCast>
             <[T; 1073741824] as shared_memory::cast::SharedMemCast>
             <[T; 10] as shared_memory::cast::SharedMemCast>
           and 55 others
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct<&[i32]>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `NotShared: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:35:5
   |
35 |     assert_impl::<MyStruct2<NotShared>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `NotShared`
   |
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct2<NotShared>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `std::vec::Vec<i32>: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:36:5
   |
36 |     assert_impl::<MyStruct2<Vec<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `std::vec::Vec<i32>`
   |
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct2<std::vec::Vec<i32>>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `&[i32]: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:37:5
   |
37 |     assert_impl::<MyStruct2<&[i32]>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `&[i32]`
   |
   = help: the following implementations were found:
             <[T; 1024] as shared_memory::cast::SharedMemCast>
             <[T; 1048576] as shared_memory::cast::SharedMemCast>
             <[T; 1073741824] as shared_memory::cast::SharedMemCast>
             <[T; 10] as shared_memory::cast::SharedMemCast>
           and 55 others
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct2<&[i32]>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `NotShared: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:42:5
   |
42 |     assert_impl::<MyData<NotShared>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `NotShared`
   |
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyData<NotShared>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `std::vec::Vec<i32>: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:43:5
   |
43 |     assert_impl::<MyData<Vec<i32>>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `std::vec::Vec<i32>`
   |
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyData<std::vec::Vec<i32>>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `&[i32]: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:44:5
   |
44 |     assert_impl::<MyData<&[i32]>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `&[i32]`
   |
   = help: the following implementations were found:
             <[T; 1024] as shared_memory::cast::SharedMemCast>
             <[T; 1048576] as shared_memory::cast::SharedMemCast>
             <[T; 1073741824] as shared_memory::cast::SharedMemCast>
             <[T; 10] as shared_memory::cast::SharedMemCast>
           and 55 others
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyData<&[i32]>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `U: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:32:9
   |
32 |         assert_impl::<MyStruct<U>>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `U`
   |
   = help: consider adding a `where U: shared_memory::cast::SharedMemCast` bound
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct<U>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `U: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:39:9
   |
39 |         assert_impl::<MyStruct2<U>>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `U`
   |
   = help: consider adding a `where U: shared_memory::cast::SharedMemCast` bound
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyStruct2<U>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `U: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/generic.rs:46:9
   |
46 |         assert_impl::<MyData<U>>();
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `U`
   |
   = help: consider adding a `where U: shared_memory::cast::SharedMemCast` bound
   = note: required because of the requirements on the impl of `shared_memory::cast::SharedMemCast` for `MyData<U>`
note: required by `assert_impl`
  --> $DIR/generic.rs:26:1
   |
26 | fn assert_impl<T: SharedMemCast>() {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0277`.
